server/src/websocket_uploader/engine_uploader.c

```diff
@@ -0,0 +1,227 @@
+ *
+ * swtbahn-cli is a RESEARCH PROTOTYPE and distributed WITHOUT ANY WARRANTY, without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ * PURPOSE. See the GNU General Public License for more details.
+ *
+ * The following people contributed to the conception and realization of the
+ * present swtbahn-cli (in alphabetic order by surname):
+ *
+ * - Bernhard Luedtke <https://github.com/bluedtke>
+ *
+ */
+
+#include "engine_uploader.h"
+#include "../server.h"
+
+#include <glib.h>
+#include "mongoose.h"
+
+typedef struct {
+	bool started;
+	bool finished;
+	bool success;
+	GString* file_path;
+	GString* srv_result_full_msg;
+} ws_verif_data;
+
+static const unsigned int websocket_single_poll_length_ms = 250;
+static const unsigned int websocket_max_polls_before_start = 60;
+
+static const char verifier_url[] = "ws://141.13.106.29:8080/engineverification/";
+// For development, local verification server.
+//static const char verifier_url[] = "ws://127.0.0.1:8080/engineverification/";
+
+static const char msg_type_field_key[] = "\"__MESSAGE_TYPE__\"";
+static const char msg_type_start_sig[] = "\"__MESSAGE_TYPE__\":\"ENG_VERIFICATION_REQUEST_START\"";
+static const char msg_type_start_sctx_field_key[] = "\"sctx\"";
+static const char msg_type_received_sig[] = "\"__MESSAGE_TYPE__\":\"ENG_VERIFICATION_REQUEST_RECEIVED\"";
+static const char msg_type_result_sig[] = "\"__MESSAGE_TYPE__\":\"ENG_VERIFICATION_REQUEST_RESULT\"";
+static const char msg_type_result_status_true_sig[] = "\"status\":true";
+static const char msg_type_result_status_false_sig[] = "\"status\":false";
+
+void process_verif_server_reply(struct mg_connection *c, struct mg_ws_message *wm, ws_verif_data* ws_data_ptr);
+void send_verif_req_message(struct mg_connection *c, ws_verif_data* ws_data_ptr);
+void websock_verification_callback(struct mg_connection *c, int ev, void *ev_data, void *fn_data);
+
+
+void send_verif_req_message(struct mg_connection *c, ws_verif_data* ws_data_ptr) {
+	// Read sctx model from file and send that as the message. 
+	// Open the file with the sctx model
+	FILE *fp = fopen(ws_data_ptr->file_path->str, "r");
+	GString *g_fullmsg = g_string_new("");
+	if (fp) {
+		char *buffer = NULL;
+		size_t length = 0;
+		// Adapted from https://stackoverflow.com/a/174743
+		ssize_t bytes_read_count = getdelim(&buffer, &length, '\0', fp);
+		if (bytes_read_count != -1 && buffer != NULL) {
+			//- Create escaped version of the GString that represents the model file
+			gchar *g_c_escaped = g_strescape(buffer, NULL);
+			//- Create message in the (json)syntax that the verification server wants
+			g_string_append_printf(g_fullmsg, "{%s,", msg_type_start_sig);
+			g_string_append_printf(g_fullmsg, "%s:\"%s\"}", msg_type_start_sctx_field_key, g_c_escaped);
+			free(g_c_escaped);
+			free(buffer);
+		} else {
+			syslog_server(LOG_ERR, "Upload Engine - Unable to read model file to be uploaded");
```

Comment:
`Websocket upload engine`?

Reply:
I would prefer "WS" instead of Websocket to reduce log length. The long lines complicate the window arrangement on the Pi